File: runtime/CSharp/Antlr3.Runtime/Antlr.Runtime.Debug/DebugTokenStream.cs
namespace Antlr.Runtime.Debug
{
    using Antlr.Runtime;
    using System;

    public class DebugTokenStream : ITokenStream, IIntStream
    {
        protected internal IDebugEventListener dbg;
        protected internal bool initialStreamState = true;
        public ITokenStream input;
        protected int lastMarker;

        public DebugTokenStream(ITokenStream input, IDebugEventListener dbg)
        {
            this.input = input;
            this.DebugListener = dbg;
            input.LT(1);
        }

        public virtual void Consume()
        {
            if (this.initialStreamState)
            {
                this.ConsumeInitialHiddenTokens();
            }
            int num = this.input.Index();
            IToken t = this.input.LT(1);
            this.input.Consume();
            int num2 = this.input.Index();
            this.dbg.ConsumeToken(t);
            if (num2 > (num + 1))
            {
                for (int i = num + 1; i < num2; i++)
                {
                    this.dbg.ConsumeHiddenToken(this.input.Get(i));
                }
            }
        }

        protected internal virtual void ConsumeInitialHiddenTokens()
        {
            int num = this.input.Index();
            for (int i = 0; i < num; i++)
            {
                this.dbg.ConsumeHiddenToken(this.input.Get(i));
            }
            this.initialStreamState = false;
        }

        public virtual IToken Get(int i)
        {
            return this.input.Get(i);
        }

        public virtual int Index()
        {
            return this.input.Index();
        }

        public virtual int LA(int i)
        {
            if (this.initialStreamState)
            {
                this.ConsumeInitialHiddenTokens();
            }
            this.dbg.LT(i, this.input.LT(i));
            return this.input.LA(i);
        }

        public virtual IToken LT(int i)
        {
            if (this.initialStreamState)
            {
                this.ConsumeInitialHiddenTokens();
            }
            this.dbg.LT(i, this.input.LT(i));
            return this.input.LT(i);
        }

        public virtual int Mark()
        {
            this.lastMarker = this.input.Mark();
            this.dbg.Mark(this.lastMarker);
            return this.lastMarker;
        }

        public virtual void Release(int marker)
        {
        }

        public virtual void Rewind()
        {
            this.dbg.Rewind();
            this.input.Rewind(this.lastMarker);
        }

        public virtual void Rewind(int marker)
        {
            this.dbg.Rewind(marker);
            this.input.Rewind(marker);
        }

        public virtual void Seek(int index)
        {
            this.input.Seek(index);
        }

        public virtual int Size()
        {
            return this.input.Size();
        }

        public override string ToString()
        {
            return this.input.ToString();
        }

        public virtual string ToString(IToken start, IToken stop)
        {
            return this.input.ToString(start, stop);
        }

        public virtual string ToString(int start, int stop)
        {
            return this.input.ToString(start, stop);
        }

        public virtual IDebugEventListener DebugListener
        {
            set
            {
                this.dbg = value;
            }
        }

        public virtual ITokenSource TokenSource
        {
            get
            {
                return this.input.TokenSource;
            }
        }
    }
}

File: src/org/antlr/runtime/debug/DebugTokenStream.java
/*
 [The "BSD licence"]
 Copyright (c) 2004 Terence Parr
 All rights reserved.

 Redistribution and use in source and binary forms, with or without
 modification, are permitted provided that the following conditions
 are met:
 1. Redistributions of source code must retain the above copyright
    notice, this list of conditions and the following disclaimer.
 2. Redistributions in binary form must reproduce the above copyright
    notice, this list of conditions and the following disclaimer in the
    documentation and/or other materials provided with the distribution.
 3. The name of the author may not be used to endorse or promote products
    derived from this software without specific prior written permission.

 THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
 IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
 OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
 IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
 INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
 NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
*/
package org.antlr.runtime.debug;

import org.antlr.runtime.*;

public class DebugTokenStream implements TokenStream {
	protected DebugEventListener dbg;
	protected TokenStream input;
	protected boolean firstConsume = true;

	public DebugTokenStream(TokenStream input, DebugEventListener dbg) {
		this.input = input;
		setDebugListener(dbg);
	}

	public void setDebugListener(DebugEventListener dbg) {
		this.dbg = dbg;
	}

	public void consume() {
		if ( firstConsume ) {
			// consume all initial off channel tokens
			int firstOnChannelTokenIndex = input.index();
			for (int i=0; i<firstOnChannelTokenIndex; i++) {
				dbg.consumeHiddenToken(input.get(i));
			}
			firstConsume = false;
		}
		int a = input.index();
		Token t = input.LT(1);
		input.consume();
		int b = input.index();
		dbg.consumeToken(t);
		if ( b>a+1 ) {
			// then we consumed more than one token; must be off channel tokens
			for (int i=a+1; i<b; i++) {
				dbg.consumeHiddenToken(input.get(i));
			}
		}
	}

	public int LA(int i) {
		dbg.LT(i, input.LT(i));
		return input.LA(i);
	}

	public Token get(int i) {
		return input.get(i);
	}

	public int mark() {
		int m = input.index();
		dbg.mark(m);
		return input.mark();
	}

	public int index() {
		return input.index();
	}

	public void rewind(int marker) {
		dbg.rewind(marker);
		input.rewind(marker);
	}

	public int size() {
		return input.size();
	}

	public Token LT(int i) {
		dbg.LT(i, input.LT(i));
		return input.LT(i);
	}

	public TokenSource getTokenSource() {
		return input.getTokenSource();
	}
}
